### PR TITLE
WIP: webhook admission: proper non-500 errors

### PR DESCRIPTION
**Pull request by sttts**
Any admission error (even simple admission denies) are returned as 500 error. This is wrong.

This PR passes through 403 and 500 error.

Are there other error we want to pass through?

```release-note
Webhook admission 403 and 500 errors are pass through to the caller of the apiserver.
```

### CODE REVIEW

**Comment by liggitt** [comment]
> Any admission error (even simple admission denies) are returned as 500 error. This is wrong.

Are you referring to an http error or an AdmissionReview containing a response.status?

If you are referring to http errors, I would not expect a 401 or 403 http response to surface as an unauthorized or forbidden error to the user. I think the inability of the apiserver to call an admission webhook should be an internal server error.

**Comment by sttts** [comment]
> Are you referring to an http error or an AdmissionReview containing a response.status?

http status code.
 
> If you are referring to http errors, I would not expect a 401 or 403 http response to surface as an unauthorized or forbidden error to the user. I think the inability of the apiserver to call an admission webhook should be an internal server error.

Mainly talking about an admission deny. It ends up as a 500 for the user right now, and should be 403. A 401 or 403 or 500 from the webhook should end up as 500 for the user.

**Comment by liggitt** [comment]
> A 401 or 403 or 500 from the webhook should end up as 500 for the user.

it's still a little unclear what this is addressing... the code seemed to pass through the status as-is previously. is the issue when the webhook returns a status that doesn't have `code`, `reason`, `details`, and `status` set coherently?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 1]
```diff
@@ -62,21 +62,29 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr *generic.Version
...
+				glog.Warningf("Failed calling webhook, failing open %v: %v", hook.Name, err)
+				utilruntime.HandleError(err)
 				continue
 			}
 			glog.Warningf("Failed calling webhook, failing closed %v: %v", hook.Name, err)
+		default:
+			glog.Warningf("Internal webhook calling error %q: %#v", hook.Name, err)
+			return apierrors.NewInternalError(err)
 		}
-		return apierrors.NewInternalError(err)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: ah, this is the important line

**Comment by DenverJ** [comment]
@sttts you mention that this PR passes through 403 and 500 but from what I can tell it passes through any valid status object returned by the webhook and sets a 403 status code by default if the webhook does not return a status code?

For me this is the desired result and just wanted to make sure I was understanding it correctly. I would like to use a StatusReason of 'Invalid' (422). If I am correct and this PR supports returning any valid status object perhaps a test case could be added to reflect that?

**Comment by sttts** [comment]
> you mention that this PR passes through 403 and 500 but from what I can tell it passes through any valid status object returned by the webhook and sets a 403 status code by default if the webhook does not return a status code?

Yes, I think you are right. Still WIP :-) Just wanted to make sure we agree on the desired behaviour (@liggitt wdyt?)

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/plugin.go [hunk 2]
```diff
@@ -92,5 +92,11 @@ func (a *Plugin) ValidateInitialization() error {
 
 // Admit makes an admission decision based on the request attributes.
 func (a *Plugin) Admit(attr admission.Attributes) error {
-	return a.Webhook.Dispatch(attr)
+	if !a.WaitForReady() {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/plugin.go]: Why moving this check from `Dispatch` to here? The old ways saves some duplicate code.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/72751 was significantly simpler and mirrored what the validating webhook was doing. Is there a reason that approach is insufficient?

**Comment by liggitt** [comment]
ping on this (and on whether https://github.com/kubernetes/kubernetes/pull/72751 is a reasonable and simpler alternative)

**Comment by sttts** [comment]
@liggitt commented on #72751. I am fine with it if we constraint it to those error we actually want to pass back to the caller.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/72751 merged and resolves this

/close

**Pull request closed without merging**